File: runtime/browser/android/xwalk_autofill_client_android.cc
// Copyright 2014 The Chromium Authors. All rights reserved.
// Copyright (c) 2015 Intel Corporation. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "xwalk/runtime/browser/android/xwalk_autofill_client_android.h"

#include "base/android/jni_android.h"
#include "base/android/jni_string.h"
#include "base/android/scoped_java_ref.h"
#include "components/autofill/core/browser/autofill_manager.h"
#include "jni/XWalkAutofillClientAndroid_jni.h"
#include "ui/gfx/geometry/rect_f.h"
#include "xwalk/runtime/browser/android/xwalk_content.h"

using base::android::AttachCurrentThread;
using base::android::ConvertUTF16ToJavaString;
using base::android::ScopedJavaLocalRef;
using content::WebContents;

DEFINE_WEB_CONTENTS_USER_DATA_KEY(xwalk::XWalkAutofillClientAndroid);

namespace xwalk {

// Ownership: The native object is created (if autofill enabled) and owned by
// XWalkContent. The native object creates the java peer which handles most
// autofill functionality at the java side. The java peer is owned by Java
// XWalkContent. The native object only maintains a weak ref to it.
XWalkAutofillClientAndroid::XWalkAutofillClientAndroid(WebContents* contents)
  : XWalkAutofillClient(contents) {
  JNIEnv* env = AttachCurrentThread();
  ScopedJavaLocalRef<jobject> delegate;
  delegate.Reset(
      Java_XWalkAutofillClientAndroid_create(
          env,
          reinterpret_cast<intptr_t>(this)));
  XWalkContent* xwalk_content = XWalkContent::FromWebContents(contents);
  xwalk_content->SetXWalkAutofillClient(delegate.obj());
  java_ref_ = JavaObjectWeakGlobalRef(env, delegate.obj());
}

void XWalkAutofillClientAndroid::ShowAutofillPopupImpl(
    const gfx::RectF& element_bounds,
    base::i18n::TextDirection text_direction,
    const std::vector<autofill::Suggestion>& suggestions) {
  JNIEnv* env = AttachCurrentThread();
  ScopedJavaLocalRef<jobject> obj = java_ref_.get(env);
  if (obj.is_null()) return;

  // We need an array of AutofillSuggestion.
  size_t count = suggestions.size();

  ScopedJavaLocalRef<jobjectArray> data_array =
      Java_XWalkAutofillClientAndroid_createAutofillSuggestionArray(env, count);

  for (size_t i = 0; i < count; ++i) {
    ScopedJavaLocalRef<jstring> name =
        ConvertUTF16ToJavaString(env, suggestions[i].value);
    ScopedJavaLocalRef<jstring> label =
        ConvertUTF16ToJavaString(env, suggestions[i].label);
    Java_XWalkAutofillClientAndroid_addToAutofillSuggestionArray(
        env, data_array.obj(), i, name.obj(), label.obj(),
        suggestions[i].frontend_id);
  }

  Java_XWalkAutofillClientAndroid_showAutofillPopup(
      env,
      obj.obj(),
      element_bounds.x(),
      element_bounds.y(),
      element_bounds.width(),
      element_bounds.height(),
      text_direction == base::i18n::RIGHT_TO_LEFT,
      data_array.obj());
}

void XWalkAutofillClientAndroid::HideAutofillPopupImpl() {
  JNIEnv* env = AttachCurrentThread();
  ScopedJavaLocalRef<jobject> obj = java_ref_.get(env);
  if (obj.is_null())
    return;
  Java_XWalkAutofillClientAndroid_hideAutofillPopup(env, obj.obj());
}

void XWalkAutofillClientAndroid::SuggestionSelected(JNIEnv* env,
                                                    jobject object,
                                                    jint position) {
  XWalkAutofillClient::SuggestionSelected(position);
}

bool RegisterXWalkAutofillClient(JNIEnv* env) {
  return RegisterNativesImpl(env);
}

}  // namespace xwalk

File: runtime/android/core_internal/src/org/xwalk/core/internal/XWalkAutofillClientAndroid.java
// Copyright 2014 The Chromium Authors. All rights reserved.
// Copyright (c) 2015 Intel Corporation. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.xwalk.core.internal;

import android.view.ViewGroup;

import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.JNINamespace;
import org.chromium.content.browser.ContentViewCore;
import org.chromium.ui.autofill.AutofillDelegate;
import org.chromium.ui.autofill.AutofillPopup;
import org.chromium.ui.autofill.AutofillSuggestion;
import org.chromium.ui.DropdownItem;

/**
 * Java counterpart to the XWalkAutofillClient. This class is owned by XWalkContent and has
 * a weak reference from native side.
 */
@JNINamespace("xwalk")
public class XWalkAutofillClientAndroid {

    private final long mNativeXWalkAutofillClientAndroid;
    private AutofillPopup mAutofillPopup;
    private ContentViewCore mContentViewCore;

    @CalledByNative
    public static XWalkAutofillClientAndroid create(long nativeClient) {
        return new XWalkAutofillClientAndroid(nativeClient);
    }

    private XWalkAutofillClientAndroid(long nativeXWalkAutofillClient) {
        mNativeXWalkAutofillClientAndroid = nativeXWalkAutofillClient;
    }

    public void init(ContentViewCore contentViewCore) {
        mContentViewCore = contentViewCore;
    }

    @CalledByNative
    private void showAutofillPopup(float x, float y, float width, float height,
            boolean isRtl, AutofillSuggestion[] suggestions) {

        if (mContentViewCore == null) return;

        if (mAutofillPopup == null) {
            mAutofillPopup = new AutofillPopup(
                mContentViewCore.getContext(),
                mContentViewCore.getViewAndroidDelegate(),
                new AutofillDelegate() {
                    @Override
                    public void dismissed() { }
                    @Override
                    public void suggestionSelected(int listIndex) {
                        nativeSuggestionSelected(mNativeXWalkAutofillClientAndroid, listIndex);
                    }
                    @Override
                    public void deleteSuggestion(int listIndex) { }
                });
        }
        mAutofillPopup.setAnchorRect(x, y, width, height);
        mAutofillPopup.filterAndShow(suggestions, isRtl);
    }

    @CalledByNative
    public void hideAutofillPopup() {
        if (mAutofillPopup == null) return;
        mAutofillPopup.dismiss();
        mAutofillPopup = null;
    }

    @CalledByNative
    private static AutofillSuggestion[] createAutofillSuggestionArray(int size) {
        return new AutofillSuggestion[size];
    }

    /**
     * @param array AutofillSuggestion array that should get a new suggestion added.
     * @param index Index in the array where to place a new suggestion.
     * @param name Name of the suggestion.
     * @param label Label of the suggestion.
     * @param uniqueId Unique suggestion id.
     */
    @CalledByNative
    private static void addToAutofillSuggestionArray(AutofillSuggestion[] array, int index,
            String name, String label, int uniqueId) {
        array[index] = new AutofillSuggestion(name, label, DropdownItem.NO_ICON, uniqueId, false, false);
    }

    private native void nativeSuggestionSelected(long nativeXWalkAutofillClientAndroid,
            int position);
}
